hw/xbox/nv2a/nv2a_pgraph.c

```diff
@@ -3041,6 +3059,40 @@ static void pgraph_bind_shaders(PGRAPHState *pg)
         }
     }
 
+    /* Window clip */
+    assert(!state.psh.window_clip_exclusive); /* FIXME: Untested */
+    state.psh.window_clip_count = 0;
+
+    /* Create region which covers the entire surface */
+    uint32_t all_x = 0;
+    uint32_t all_y = 0;
+    SET_MASK(all_x, NV_PGRAPH_WINDOWCLIPX0_XMIN, 0);
+    SET_MASK(all_x, NV_PGRAPH_WINDOWCLIPX0_XMAX,
+                    pg->surface_shape.clip_width - 1);
+    SET_MASK(all_y, NV_PGRAPH_WINDOWCLIPX0_XMIN, 0);
+    SET_MASK(all_y, NV_PGRAPH_WINDOWCLIPX0_XMAX,
+                    pg->surface_shape.clip_height - 1);
+    uint32_t last_x = all_x;
+    uint32_t last_y = all_y;
+
+    /* Check if the last region is duplicated and ignore duplicates. We'll also
+     * early out on regions which cover the entire surface.
+     */
+    for (i = 0; i < 8; i++) {
+        uint32_t x = pg->regs[NV_PGRAPH_WINDOWCLIPX0 + i * 4];
+        uint32_t y = pg->regs[NV_PGRAPH_WINDOWCLIPY0 + i * 4];
+        /* Region covers entire surface = other regions will be overlap */
+        if ((x == all_x) && (y == all_y)) {
+            state.psh.window_clip_count = 0;
+            break;
+        }
+        if (x != last_x || y != last_y) {
+            state.psh.window_clip_count = i + 1;
+            last_x = x;
+            last_y = y;
+        }
+    }
```

Comment:
nit: Please review this optimization logic. I'm a little worried about it to be honest.
In particular the `all_x`, `all_y` part. the `last_x` / `last_y` should work fine.

I remember this stuff making my brain hurt (especially as window clip is not really documented anywhere and it's hard to find good tests)

Reply:
I'm pretty sure the logic is fine and fast enough. It might be fine just to let the shader do all 8 clipping regions. This definitely won't simplify everything, but it'll catch duplicates at least.

Reply:
The results in Halo imply that `max` bounds are inclusive. So `(0,0)-(0,0)` would still clip a single pixel. This needs more testing on real hardware though (also with anti-aliasing). We can't test in XQEMU because our pixel centers are wrong (I think), so we get off-by-half and off-by-one errors anyway.
It's up to you how you implement that case, we should have an issue about this ("Clarify clip region bounds" or something).

I believe the idea behind the `all` was to be able to avoid some draw call altogethers, and avoid *any* use of `discard`, because that could mean taking a slow path in the rasterizer of some older GPUs. It's probably neglible though, so I'm fine with this change; especially because it mostly adds complexity.

The  `last_x` and `last_y` should probably remain. I believe Dead or Alive 3 repeated clip regions. Not sure if this has any impact, but it's such a simple optimzation that it can be kept - probably more as an inspiration for better optimizations or if we move the clipping to the VSH (clipplanes or something).
But I'll also leave this up to you.